### PR TITLE
Added calculation of the distance between a line and a point

### PR DESCRIPTION
**Pull request by Darya177777**

### CODE REVIEW

**Comment by esabol** [comment]
CI tests failed. Make sure your fork is sync'ed with this repo. You might need to just update the line numbers in some expected output files?

**Comment by Darya177777** [comment]
@esabol : All tests fixed. Please, re-review.

File: doc/operators.sgm [hunk 1]
```diff
@@ -331,7 +331,7 @@
                 a non-boolean operator returning the distance between two
                 objects in radians. Currently,
                 <application>pgSphere</application> supports only distances
-                between points, circles, and between point and circle. If the
+                between points, circles, between point and circle, and between point and line. If the
```
Comment (esabol) [inline doc/operators.sgm:334]: Does this line exceed 79 characters? If so, please word-wrap.

File: pgs_line.sql.in [hunk 2]
```diff
@@ -593,3 +593,53 @@ COMMENT ON FUNCTION scircle_contains_line_com_neg(sline, scircle) IS
...
+--
+
+CREATE FUNCTION dist(sline, spoint)
+   RETURNS FLOAT8
+   AS 'MODULE_PATHNAME', 'sphereline_point_distance'
+   LANGUAGE 'c'
+   IMMUTABLE STRICT;
+
+COMMENT ON FUNCTION dist(sline, spoint) IS
+  'distance between spherical line and spherical point';
```
Comment (esabol) [inline pgs_line.sql.in:607]: All the other COMMENT ON FUNCTIONs in this file start with "returns..." I know that seems redundant, but I think it's best to fit in with the convention of the other COMMENTs.

File: expected/line.out [hunk 3]
```diff
@@ -134,4 +134,34 @@ SELECT sline ( spoint '(0,   0d)', spoint '(0.000001d, 0d)' )   #
  f
 (1 row)
 
-       
+-- checking the distance between a line and a point
+SELECT round( CAST(sline '( 90d, 0d, 0d, XYZ ), 40d ' <-> spoint '( 0d, 90d )' as numeric), 14);
```
Comment (esabol) [inline expected/line.out:138]: What about the distance between a point and a line? I.e., the commutative property. Is that implemented? If not, it should be. If it is, then it should be tested.
```
SELECT round( CAST(spoint '( 0d, 90d )' <-> sline '( 90d, 0d, 0d, XYZ ), 40d ' as numeric), 14);
```

**Review by esabol — CHANGES_REQUESTED** [review]
Thank you for this! Looks pretty good to me. Just some minor changes, I think.

**Comment by Darya177777** [comment]
All fixed.

**Comment by esabol** [comment]
> All fixed.

OK, looking better! @vitcpp will probably have some comments as well.

**Comment by vitcpp** [comment]
@Darya177777 Could you also please rebase on the newest version and update the upgrade script upgrade_scripts/pg_sphere--1.2.3--1.3.0.sql.in. Such script is used when updating the extension to a newer version using ALTER EXTENSION UPDATE TO command. Once you have introduced a new function the upgrade script should be updated.

File: src/line.c [hunk 4]
```diff
@@ -642,6 +644,59 @@ sline_center(SPoint *c, const SLine *sl)
...
+	vector3d_spoint(&fp, &first_p);
+
+	if (spoint_at_sline(&fp, sl))
+	{
+		return spoint_dist(&fp, p);
+	}
+
+	vector3d_spoint(&fp, &v_beg);
+	vector3d_spoint(&sp, &v_end);
+	if (FPle(spoint_dist(p, &fp), spoint_dist(p, &sp)))
```
Comment (vitcpp) [inline src/line.c:693]: I would propose to this code faster like this:
fpdist = spoint_dist(p, &fp);
spdist = spoint_dist(p, &sp);
return Min(fpdist, spdist);

File: src/line.c [hunk 5]
```diff
@@ -1051,3 +1106,20 @@ spheretrans_from_line(PG_FUNCTION_ARGS)
 	sphereline_to_euler(e, l);
 	PG_RETURN_POINTER(e);
 }
+
+Datum
+sphereline_point_distance(PG_FUNCTION_ARGS)
+{
+	SLine	   *s = (SLine *) PG_GETARG_POINTER(0);
```
Comment (vitcpp) [inline src/line.c:1113]: I would propose to add const prefix in local variable declarations.

File: src/line.c [hunk 6]
```diff
@@ -1051,3 +1106,20 @@ spheretrans_from_line(PG_FUNCTION_ARGS)
...
+	SPoint	   *p = (SPoint *) PG_GETARG_POINTER(1);
+	PG_RETURN_FLOAT8(sline_point_dist(s, p));
+}
+
+Datum
+sphereline_point_distance_com(PG_FUNCTION_ARGS)
+{
+	SPoint	   *p = (SPoint *) PG_GETARG_POINTER(0);
```
Comment (vitcpp) [inline src/line.c:1121]: I would propose to add const prefix in local variable declarations.

File: expected/line.out [hunk 7]
```diff
@@ -134,4 +134,35 @@ SELECT sline ( spoint '(0,   0d)', spoint '(0.000001d, 0d)' )   #
...
+      round       
+------------------
+ 0.87266462599717
+(1 row)
+
+SELECT round( CAST(sline '( 90d, 0d, 0d, XYZ ), 40d ' <-> spoint '( 0d, 90d )' as numeric), 14)   =
```
Comment (vitcpp) [inline expected/line.out:144]: I think round(CAST... is not necessary.
I would propose to add SET extra_float_digits TO -3; in the beginning of file. It reduces the number of output digits.

I've attached the modified line.sql 
[line.sql.zip](https://github.com/postgrespro/pgsphere/files/12266486/line.sql.zip)

**Review by vitcpp — CHANGES_REQUESTED** [review]

**Comment by vitcpp** [comment]
@Darya177777 Could you please squash two commits into one.

File: doc/index.html [hunk 8]
```diff
@@ -0,0 +1,227 @@
+<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN""http://www.w3.org/TR/html4/loose.dtd">
```
Comment (vitcpp) [inline doc/index.html:1]: This file is generated. It should not be added in the commit. Could you please remove it?

File: doc/types.html [hunk 9]
```diff
@@ -0,0 +1,272 @@
+<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN""http://www.w3.org/TR/html4/loose.dtd">
```
Comment (vitcpp) [inline doc/types.html:1]: This file should be removed from the commit.

File: doc/x20.html [hunk 10]
```diff
@@ -0,0 +1,243 @@
+<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN""http://www.w3.org/TR/html4/loose.dtd">
```
Comment (vitcpp) [inline doc/x20.html:1]: Remove this file from the commit.

File: doc/x46.html [hunk 11]
```diff
@@ -0,0 +1,423 @@
+<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN""http://www.w3.org/TR/html4/loose.dtd">
```
Comment (vitcpp) [inline doc/x46.html:1]: Remove this file from the commit.

File: expected/init_test.out [hunk 12]
```diff
@@ -0,0 +1,37 @@
+SET client_min_messages TO NOTICE;
```
Comment (vitcpp) [inline expected/init_test.out:1]: Remove this file from the commit.

File: log/initdb.log [hunk 13]
```diff
@@ -0,0 +1,37 @@
+Running in no-clean mode.  Mistakes will not be cleaned up.
```
Comment (vitcpp) [inline log/initdb.log:1]: Remove this file from the commit.

File: log/postmaster.log [hunk 14]
```diff
@@ -0,0 +1,376 @@
+2023-08-08 15:16:01.222 +07 postmaster[14615] LOG:  starting PostgreSQL 15.3 on x86_64-pc-linux-gnu, compiled by gcc (Ubuntu 9.4.0-1ubuntu1~20.04.1) 9.4.0, 64-bit
```
Comment (vitcpp) [inline log/postmaster.log:1]: Remove this file from the commit.

File: logfile [hunk 15]
```diff
@@ -0,0 +1,749 @@
+2023-08-08 14:58:03.794 +07 [12364] LOG:  starting PostgreSQL 15.3 on x86_64-pc-linux-gnu, compiled by gcc (Ubuntu 9.4.0-1ubuntu1~20.04.1) 9.4.0, 64-bit
```
Comment (vitcpp) [inline logfile:1]: Remove this file from the commit.

**Comment by vitcpp** [comment]
PR seems to be ok. I need to do some testing before merge.

**Comment by esabol** [comment]
> @Darya177777 Could you please squash two commits into one.

@vitcpp, doesn't GitHub automatically squash commits on merge these days? I've been under the impression that it does, but maybe I'm used to GitLab.

**Comment by vitcpp** [comment]
@esabol I'm not sure about github feature to squash commits. I will check it. I'm concerned which final comment will appear in commit after squashing. I'm not sure that simple joining of commit messages is a good idea. Furthermore, it is harder to review multiple commits where latest commit overwrites some changes of the previous commit. PR should contain the final version of the changes. It can contain multiple commits but these commits should be used to logically separate the changes. They should not touch the same code.

**Review by vitcpp — APPROVED** [review]

**Comment by vitcpp** [comment]
@Darya177777 Could you please fix the tests? Thank you.

File: src/sscan.c [hunk 16]
```diff
@@ -1,6 +1,6 @@
-#line 2 "sscan.c"
```
Comment (vitcpp) [inline src/sscan.c:1]: This file is generated. It should be removed from the commit.

**Review by vitcpp — CHANGES_REQUESTED** [review]

**Comment by vitcpp** [comment]
@Darya177777 You have added sscan.c file to the commit. It shouldn't be added. Could you please remove it from the commit.

**Comment by esabol** [comment]
> @esabol I'm not sure about github feature to squash commits. I will check it. I'm concerned which final comment will appear in commit after squashing. I'm not sure that simple joining of commit messages is a good idea. Furthermore, it is harder to review multiple commits where latest commit overwrites some changes of the previous commit. PR should contain the final version of the changes. It can contain multiple commits but these commits should be used to logically separate the changes. They should not touch the same code.

@vitcpp, you can edit the merge commit message to whatever you want when squashing and merging, according to this: https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/incorporating-changes-from-a-pull-request/about-pull-request-merges#squash-and-merge-your-commits

**Comment by vitcpp** [comment]
> @vitcpp, you can edit the merge commit message to whatever you want when squashing and merging, according to this: https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/incorporating-changes-from-a-pull-request/about-pull-request-merges#squash-and-merge-your-commits

@esabol  Thank you very much! I will take a look.

**Review by vitcpp — APPROVED** [review]

**Review by vitcpp — CHANGES_REQUESTED** [review]
@Darya177777 It seems we missed to make some changes in pg_sphere--1.2.3--1.3.0.sql.in. This script is used to upgrate pgsphere versions in the existing database. Just copy the contents of pgs_line.sql.in into this file.

**Comment by Darya177777** [comment]
@vitcpp I have made some changes in pg_sphere--1.2.3--1.3.0.sql.in. Please, re-review.

**Review by esabol — APPROVED** [review]
Looks good to me.

**Review by vitcpp — APPROVED** [review]

**Pull request merged**